merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,1211 @@
+        let indices_size = indices_size_unaligned + mem::align_of::<usize>()
+            - (indices_size_unaligned % mem::align_of::<usize>());
+
+        let values_size_unaligned = mem::size_of::<Option<HashSetCell>>() * capacity_values;
+        // Make sure that alignment of `values` matches the alignment of `usize`.
+        let values_size = values_size_unaligned + mem::align_of::<usize>()
+            - (values_size_unaligned % mem::align_of::<usize>());
+
+        Ok(dyn_fields_size + next_value_index_size + indices_size + values_size)
+    }
+
+    /// Returns the capacity of buckets for the desired `capacity`, while taking
+    /// the load factor in account.
+    pub fn capacity_indices(
+        capacity_elements: usize,
+        load_factor: f64,
+    ) -> Result<f64, HashSetError> {
+        // To treat `capacity_elements` as `f64`, we need to fit it in `u32`.
+        // `u64`/`usize` can't be casted directoy to `f64`.
+        let capacity_elements =
+            u32::try_from(capacity_elements).map_err(|_| HashSetError::IntegerOverflow)?;
+        let minimum = f64::from(capacity_elements) / load_factor;
+        Ok(find_next_prime(minimum))
+    }
+
+    // Create a new hash set with the given capacity
+    pub fn new(
+        capacity_indices: usize,
+        capacity_values: usize,
+        sequence_threshold: usize,
+    ) -> Result<Self, HashSetError> {
+        let layout = Layout::new::<usize>();
+        // SAFETY: Allocating a primitive type. There is no chance of
+        // misalignment.
+        let next_value_index = unsafe { alloc::alloc(layout) as *mut usize };
+        if next_value_index.is_null() {
+            handle_alloc_error(layout);
+        }
+        unsafe {
+            *next_value_index = 0;
+        }
+
+        let layout = Layout::array::<Option<I>>(capacity_indices).unwrap();
+        // SAFETY: `I` is always a signed integer. Creating a layout for an
+        // array of integers of any size won't cause any panic.
+        let indices_ptr = unsafe { alloc::alloc(layout) as *mut Option<I> };
+        if indices_ptr.is_null() {
+            handle_alloc_error(layout);
+        }
+        let indices = NonNull::new(indices_ptr).unwrap();
+        for i in 0..capacity_indices {
+            unsafe {
+                std::ptr::write(indices_ptr.add(i), None);
+            }
+        }
+
+        // SAFETY: `I` is always a signed integer. Creating a layout for an
+        // array of integers of any size won't cause any panic.
+        let layout = Layout::array::<Option<HashSetCell>>(capacity_values).unwrap();
+        let values_ptr = unsafe { alloc::alloc(layout) as *mut Option<HashSetCell> };
+        if values_ptr.is_null() {
+            handle_alloc_error(layout);
+        }
+        let values = NonNull::new(values_ptr).unwrap();
+        for i in 0..capacity_values {
+            unsafe {
+                std::ptr::write(values_ptr.add(i), None);
```

Comment:
why does value pointer plus integer i result in the correct position of the hash cell sequence number?
Wouldn't it be ``i *  (size of hash cell + 32)``

Reply:
See the comment above. `.add(i)` moves the pointer by `i * size_of_the_ptr_type` already.